rules_gapic/python/py_gapic.bzl

```diff
@@ -0,0 +1,170 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+load("//rules_gapic:gapic.bzl", "GapicInfo", "gapic_srcjar", "unzipped_srcjar")
+
+def _py_gapic_postprocessed_srcjar_impl(ctx):
+    gapic_srcjar = ctx.file.gapic_srcjar
+    formatter = ctx.executable.formatter
+    output_dir_name = ctx.label.name
+
+    output_main = ctx.outputs.main
+    output_test = ctx.outputs.test
+    output_pkg = ctx.outputs.pkg
+    outputs = [output_main, output_test, output_pkg]
+
+    output_dir_path = "%s/%s" % (output_main.dirname, output_dir_name)
+
+    # Note the script is more complicated than it intuitively should be because of limitations
+    # inherent to bazel execution environment: no absolute paths allowed, the generated artifacts
+    # must ensure uniqueness within a build.
+    script = """
+    unzip -q {gapic_srcjar} -d {output_dir_path}
+    {formatter} -q {output_dir_path}
+    pushd {output_dir_path}
+    zip -q -r {output_dir_name}-pkg.srcjar nox.py setup.py setup.cfg docs MANIFEST.in README.rst LICENSE
+    rm -rf nox.py setup.py docs
+    zip -q -r {output_dir_name}-test.srcjar tests
+    rm -rf tests
+    zip -q -r {output_dir_name}.srcjar . -i \*.py
+    popd
+    mv {output_dir_path}/{output_dir_name}.srcjar {output_main}
+    mv {output_dir_path}/{output_dir_name}-test.srcjar {output_test}
+    mv {output_dir_path}/{output_dir_name}-pkg.srcjar {output_pkg}
+    rm -rf {output_dir_path}
+    """.format(
+        gapic_srcjar = gapic_srcjar.path,
+        output_dir_name = output_dir_name,
+        output_dir_path = output_dir_path,
+        formatter = formatter.path,
+        output_main = output_main.path,
+        output_test = output_test.path,
+        output_pkg = output_pkg.path,
+    )
+
+    ctx.actions.run_shell(
+        inputs = [gapic_srcjar],
+        command = script,
+        outputs = outputs,
+    )
+
+    return [
+        DefaultInfo(
+            files = depset(direct = outputs)
+        ),
+        GapicInfo(
+            main = output_main,
+            test = output_test,
+            pkg = output_pkg,
+        ),
+    ]
+
+_py_gapic_postprocessed_srcjar = rule(
+    implementation = _py_gapic_postprocessed_srcjar_impl,
+    attrs = {
+        "gapic_srcjar": attr.label(
+            doc = "The srcjar of the generated GAPIC client.",
+            mandatory = True,
+            allow_single_file = True,
+        ),
+        "formatter": attr.label(
+            doc = "Formats the output to a Python-idiomatic style.",
+            default = Label("@pypi_black//:black"),
+            executable = True,
+            cfg = "host",
+        ),
+    },
+    outputs = {
+      "main": "%{name}.srcjar",
+      "test": "%{name}-test.srcjar",
```

Comment:
Sorry I should have mentioned it in first pass (when asked to add pkg). What about the `smoke-test` part? Not all clients have it, but those which do, usually require special handling. 
The reason why we want to separate unit tests and smoke tests is that we want to run unit tests but not smoke tests, because those require special environment setup and in general much harder to control than unit tests.

You can use pubsub or some other complicated clients for an example of a client which has smoke tests configured.

Reply:
Done